### PR TITLE
test: correct many tests that were leaving clients open and hanging

### PR DESCRIPTION
**Pull request by mbroadst**
NODE-2085

### CODE REVIEW

File: test/functional/causal_consistency_tests.js [hunk 1]
```diff
@@ -25,7 +24,11 @@ describe('Causal Consistency', function() {
     test.client = this.configuration.newClient({ w: 1 }, { poolSize: 1, auto_reconnect: false });
     return test.client.connect();
   });
-  afterEach(() => test.client.close());
+
+  afterEach(() => {
+    test.listener.uninstrument();
```
Comment (daprahamian) [inline test/functional/causal_consistency_tests.js]: unrelated: should we be testing with `uninstrument`?

Comment (mbroadst) [reply test/functional/causal_consistency_tests.js]: probably not, but seems orthogonal to the goal of this PR

File: test/functional/aggregation_tests.js [hunk 2]
```diff
@@ -892,8 +882,7 @@ describe('Aggregation', function() {
                       expect(err).to.be.null;
                       expect(count).to.be.greaterThan(0);
 
-                      client.close();
-                      done();
+                      client.close(done);
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/functional/aggregation_tests.js]: No, this PR does not solve that issue unfortunately, but this is an incremental step towards better test runs: when our tests succeed, all the clients are now closed after the fact. I imagine we will want another stab at this in the future to complete the job, for now I'm just trying to get mocha to run to completion without hanging indefinitely/infinitely.

File: test/functional/mongo_client_tests.js [hunk 3]
```diff
@@ -629,7 +629,7 @@ describe('MongoClient', function() {
     test: function(done) {
       var configuration = this.configuration;
       var url = configuration.url();
-      if (url.indexOf('replicaSet') !== -1) {
+      if (url.indexOf('rs_name') !== -1) {
```
Comment (daprahamian) [inline test/functional/mongo_client_tests.js]: how do we know that our url will always have `rs_name` and not `replicaSet`? Could we check for both here?

More importantly, it looks like this is specifically trying to figure out whether to append a `?` or a `&` to the url string. Could we do:

```js
if (url.indexOf('?') !== -1) {
  url = `${configuration.url}&appname=hello%20world`;
} else {
  url = `${configuration.url}?appname=hello%20world`;
}
```

Comment (mbroadst) [reply test/functional/mongo_client_tests.js]: yeah, actually this caused an error. For now I've reverted to `rs_name`, and will revert that to `replicaSet` in the new test runner branch. This is specific to how we build urls in the configuration, where the new config will build more up-to-date connection strings.

**Pull request merged**